### PR TITLE
Add Apps API page

### PR DESCRIPTION
**Pull request by rob42**
Documenting the `signalk/v1/apps` API

### CODE REVIEW

**Comment by rob42** [comment]
Freeboard-sk loads your choice of instruments in a sidepanel. It needs to be able to get name, url. I would like to see KIP do the same in reverse, so you create an MFD by inserting apps into panels.

I included search, install, remove so that they could potentially offer options that were not currently installed.

**Comment by sbender9** [comment]
Got you. Like it! And now I want to do this in WilhelmSK ;)

**Comment by sbender9** [comment]
Can we add some kind of “status” to the data from /list? Maybe with values like `available`, `installing`, ‘installation failed’?

Maybe `installed/restart required` in case the server need restarted after installation? Or maybe I just need to fix node server so restarts on required after install :(

**Comment by rob42** [comment]
That supposes the install process has a well-defined process and outcome and the wrapper can store data on that somewhere. Outside of node/npm, install is just download the archive, unpack, and search for the index.html (all over the place..). No real way to know if its ok. 

I create the `list` output by traversing the install dirs and parsing package.json. So if its there its listed. I create the `search` list by a search and then mark the installed ones, which allows me to add a 'Remove' button. But thats server specific really, adding extra data to the reply json

**Comment by rob42** [comment]
given the mess of current webapps package formats I think that needs sorting first

**Comment by panaaj** [comment]
Some added value to consider (not mandatory though) is meta-data that can allow inclusion of app icons, etc to be displayed when listing apps... for example [#629](https://github.com/SignalK/signalk-server-node/issues/629)

**Comment by rob42** [comment]
Added the manifest.json if it exists
`"manifest": {..content of manifest.json..}`

**Comment by sbender9** [comment]
@rob42 I don’t see the issue with status. Node server provides this already. Maybe make it optional in case the server can’t provide it?

**Comment by rob42** [comment]
Theres no problem sending additional data thats implementation dependent. Its especially useful for a servers app management ui. When the install is just unzip, if its there status=INSTALLED, there is no FAILED state,  just a failed install, eg there is no way to tell its installed but not working. So I dont see how it would be useful.

**Comment by rob42** [comment]
> Maybe installed/restart
Since a webapp is just a bunch of static files  a reboot seems a bit extreme 😮

**Comment by sbender9** [comment]
Lots of things could fail when installing one of these. No internet. Missing dependencies. No disk space.

I think letting the user know that the install failed would be very useful. And like we do with node server, actually show the user information as to why it failed would be great too. 

I think providing a way for users to install web apps and not giving them any way to see the progress and results of the install is short sighted.

**Comment by rob42** [comment]
I think you misunderstand me. Definitely things can fail in the install, and showing progress is useful, but I see that as part of the install process, which is implementation dependant. This api is just to find installed apps, search for apps, and being able to request an install or remove is just a conveinience

How would you structure feedback from the install?

**Comment by rob42** [comment]
@sbender9 ?

**Comment by fabdrol** [comment]
Would it be out of scope to add to this that a server _may_ provide the apps via the various HTML5 discovery protocols of the MFDs (i.e. Garmin OneHelm & Navico's implementation)?

File: gitbook-docs/apps_api.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+## List
+
+`signalk/v1/apps/list` lists the currently install webapps. It is typically used to create app menus, and manage installed app lists.
+It returns a json array in the form:
+```
+[{
+//mandatory
+	"name": "Kip",
+	"description": "Really cool instrument app!!",
+	"location": "../@mxtommy/kip/package/public/index.html",
```
Comment (fabdrol) [inline gitbook-docs/apps_api.md]: @rob42 when implementing a client I'd prefer an absolute path from the root of the server. @sbender9 your thoughts from Wilhelm SK perspective?

File: gitbook-docs/apps_api.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+....
+]
+```
+
+## Search
+
+`signalk/v1/apps/search?keyword=signalk-webapp` finds the currently available webapps from npm. It is typically used to list available apps for installation.
+By default it makes a call to `https://api.npms.io/v2/search?size=250&q=keywords:signalk-webapp`
+It returns a json array in the form:
+```
```
Comment (fabdrol) [inline gitbook-docs/apps_api.md]: Shouldn't we include pagination here as part of the search request? I.e. `signalk/v1/apps/search?keyword=foo&skip=0&limit=25`? If things grow that may be necessary

File: gitbook-docs/apps_api.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+
+# Install
+
+`signalk/v1/apps/install?appName=@signalk/freeboard-sk&appVersion=1.3.1`, with params selected from above. This will install the app into
+the preferred location on the Signal K server, from which the `list` call can be used to generate the app menu.
+
+# Remove
+
+`signalk/v1/apps/remove?appName=@signalk/freeboard-sk`, with params selected from `list`. This will remove the app.
```
Comment (fabdrol) [inline gitbook-docs/apps_api.md]: `uninstall` makes more sense than remove (as we use `install` rather than `add`)

**Comment by rob42** [comment]
> Would it be out of scope to add to this that a server _may_ provide the apps via the various HTML5 discovery protocols of the MFDs (i.e. Garmin OneHelm & Navico's implementation)?

I think that falls under implementation. The implementation is free to do whatever, this is about  _signalk_ webapps.

File: gitbook-docs/apps_api.md (hunk 1 above)
Comment (rob42) [reply gitbook-docs/apps_api.md]: done

File: gitbook-docs/apps_api.md (hunk 2 above)
Comment (rob42) [reply gitbook-docs/apps_api.md]: Lets add that when we need it. If we add pagination etc it should be generic to all the signalk API's. That includes pagination, ordering,  search hits, max size, lots of stuff etc

**Comment by fabdrol** [comment]
> I think that falls under implementation

@rob42 I agree that the _how_ falls under implementation, but way could promote this ourselves. This is an area where Signal K could really come to its own...

**Comment by tkurki** [comment]
I agree with Rob. 

I don't quite see how saying that a server may support manufacturers' propietary system X (where each manufacturer has their own prop way of doing things) is useful, even if X is something that Signal K does / could do well.

A server may support all kinds of things, the spec should concentrate on Signal K.

**Comment by fabdrol** [comment]
Okay, in that case shall we merge?

**Comment by rob42** [comment]
Yes

**Pull request merged**